modules/puppetca/puppetca_main.rb

```diff
@@ -16,51 +18,48 @@ def clean certname
       puppetca("clean", certname)
     end
 
-    #remove certname from autosign if exists
-    def disable certname
-      raise "No such file #{autosign_file}" unless File.exist?(autosign_file)
-
-      found = false
-      entries = File.readlines(autosign_file).collect do |l|
-        if l.chomp != certname
-          l
-        else
-          found = true
-          nil
-        end
-      end.uniq.compact
-      if found
-        open(autosign_file, File::TRUNC|File::RDWR) do |autosign|
-          autosign.write entries.join
-        end
-        logger.debug "Removed #{certname} from autosign"
-      else
-        logger.debug "Attempt to remove nonexistent client autosign for #{certname}"
-        raise NotPresent, "Attempt to remove nonexistent client autosign for #{certname}"
+    # decide if csr should be autosigned
+    # parameter is csr to use
+    def autosign csr
+      if csr.nil?
+        logger.debug "Request did not include a CSR"
+        return false
+      end
+      return true if signall
+      begin
+        req = Proxy::PuppetCa::CertificateRequest.new csr
+      rescue OpenSSL::X509::RequestError
+        logger.debug "Invalid CSR"
+        return false
       end
+      token = req.challenge_password
+      if token.nil?
+        logger.debug "CSR did not include a token"
+        return false
+      end
+      foreman_callback token
     end
 
-    # add certname to puppet autosign file
-    # parameter is certname to use
-    def autosign certname
-      FileUtils.touch(autosign_file) unless File.exist?(autosign_file)
-
-      open(autosign_file, File::RDWR) do |autosign|
-        # Check that we don't have that host already
-        found = autosign.readlines.find { |line| line.chomp == certname }
-        autosign.puts certname unless found
-      end
-      logger.debug "Added #{certname} to autosign"
+    def signall
+      Proxy::PuppetCa::Plugin.settings.signall
     end
 
-    # list of hosts which are now allowed to be installed via autosign
-    def autosign_list
-      return [] unless File.exist?(autosign_file)
-      File.read(autosign_file).split("\n").reject do |v|
-        v =~ /^\s*#.*|^$/ ## Remove comments and empty lines
-      end.map do |v|
-        v.chomp ## Strip trailing spaces
+    def foreman_callback token
+      logger.debug "Calling foreman with token '#{token}'"
+      fr = Proxy::HttpRequest::ForemanRequest.new
+      req = fr.request_factory.create_delete("api/puppetca_token", :id => token)
+      begin
+        response = fr.send_request(req)
+        if response.code.to_i == 204
+          logger.debug "Autosigning CSR with token " + token
+          return true
+        end
+      rescue => e
+        logger.debug "Error while calling foreman: " + e.to_s
```

Comment:
Please log this at the error level.

Reply:
Will change that. Do you think that the other logs are correct on debug? Considering you normally don't run it on debug-loglevel, you won't be able to trace back why a CSR wasn't verified if you wanted.

Reply:
I'm ok either way, but perhaps a warning would be a bit more helpful.